lib/ex_unit/lib/ex_unit/case.ex

```diff
@@ -532,15 +550,109 @@ defmodule ExUnit.Case do
+    do_register_attribute(:ex_unit_registered_test_attributes, env, name, opts)
+  end
+
+  @doc """
+  Registers a new describe attribute to be used during `ExUnit.Case` tests.
+
+  The attribute values will be available through `context.registered`.
+  Registered values are cleared after each `ExUnit.Case.describe/2` similar
+  to `@describetag`.
+
+  `Module.register_attribute/3` is used to register the attribute,
+  this function takes the same options.
+
+  ## Examples
+
+      defmodule MyTest do
+        use ExUnit.Case
+
+        ExUnit.Case.register_describe_attribute(__MODULE__, :describe_fixtures, accumulate: true)
+
+        describe "using custom attribute" do
+          @describe_fixtures :user
+          @describe_fixtures {:post, insert: false}
+
+          test "has attribute", context do
+            assert context.registered.describe_fixtures == [{:post, insert: false}, :user]
+          end
+        end
+
+        describe "custom attributes are cleared per describe" do
+          test "doesn't have attributes", context do
+            assert context.registered.describe_fixtures == []
+          end
+        end
+      end
+
+  """
+  @spec register_describe_attribute(env, name :: atom, opts :: keyword) :: :ok
+  def register_describe_attribute(env, name, opts \\ []) do
+    do_register_attribute(:ex_unit_registered_describe_attributes, env, name, opts)
+  end
+
+  @doc """
+  Registers a new module attribute to be used during `ExUnit.Case` tests.
+
+  The attribute values will be available through `context.registered`.
+
+  `Module.register_attribute/3` is used to register the attribute,
+  this function takes the same options.
+
+  ## Examples
+
+      defmodule MyTest do
+        use ExUnit.Case
+
+        ExUnit.Case.register_module_attribute(__MODULE__, :module_fixtures, accumulate: true)
+
+        @module_fixtures :user
+        @module_fixtures {:post, insert: false}
+
+        test "using custom attribute", context do
+          assert context.registered.fixtures == [{:post, insert: false}, :user]
+        end
+
+        test "still using custom attribute", context do
+          assert context.registered.fixtures == [{:post, insert: false}, :user]
+        end
+      end
+
+  """
+  @spec register_module_attribute(env, name :: atom, opts :: keyword) :: :ok
+  def register_module_attribute(env, name, opts \\ []) do
+    do_register_attribute(:ex_unit_registered_module_attributes, env, name, opts)
+  end
 
-  def register_attribute(%{module: module}, name, opts) do
-    register_attribute(module, name, opts)
+  defp do_register_attribute(type, %{module: module}, name, opts) do
+    do_register_attribute(type, module, name, opts)
   end
 
-  def register_attribute(mod, name, opts) when is_atom(mod) and is_atom(name) and is_list(opts) do
+  defp do_register_attribute(type, mod, name, opts)
+       when is_atom(type) and is_atom(mod) and is_atom(name) and is_list(opts) do
```

Comment:
I think it would be better to move the guards to the public `register_module_attribute/3` so that if a `FunctionClauseError` happens it will point to the public function.

Reply:
I can pull up the others, but the `is_atom(mod)` is hard to pull out here, since the public function allows an ENV or a module. Recommendations for that?

Reply:
If I leave it off, it get's cause by Module.register_attribute?/3. Maybe that's good enough?

```
** (FunctionClauseError) no function clause matching in Module.register_attribute/3

    The following arguments were given to Module.register_attribute/3:

        # 1
        1

        # 2
        :foo

        # 3
        []

    Attempted function clauses (showing 1 out of 1):

        def register_attribute(module, attribute, options) when is_atom(module) and is_atom(attribute) and is_list(options)

    (elixir) lib/module.ex:1249: Module.register_attribute/3
    iex:7: (module)
```

Reply:
If `mod` is not an atom isn't it a very weird error? I'm fine with that if that's the case. We tend to not use defensive guards in private functions.

Reply:
Yes, we should move the guards up for nice funcction clause errors as smuch as possible.

Reply:
This is addressed